server/datastore/mysql/software.go

```diff
@@ -252,6 +272,45 @@ func insertNewInstalledHostSoftwareDB(
 	return nil
 }
 
+// update host_software when incoming software has a significantly more recent
+// last opened timestamp (or didn't have on in currentMap). Note that it only
+// processes software that is in both current and incoming maps, as the case
+// where it is only in incoming is already handled by
+// insertNewInstalledHostSoftwareDB.
+func updateModifiedHostSoftwareDB(
+	ctx context.Context,
+	tx sqlx.ExtContext,
+	hostID uint,
+	currentMap map[string]fleet.Software,
+	incomingMap map[string]fleet.Software,
+) error {
+	const stmt = `UPDATE host_software SET last_opened_at = ? WHERE host_id = ? AND software_id = ?`
+
+	var keysToUpdate []string
+	for key, newSw := range incomingMap {
+		curSw, ok := currentMap[key]
+		if !ok || newSw.LastOpenedAt == nil {
+			// software must also exist in current map, and new software must have a
+			// last opened at timestamp (otherwise we don't overwrite the old one)
+			continue
+		}
+
+		if curSw.LastOpenedAt == nil || (*newSw.LastOpenedAt).Sub(*curSw.LastOpenedAt) >= minLastOpenedAtDiff {
+			keysToUpdate = append(keysToUpdate, key)
+		}
+	}
+	sort.Strings(keysToUpdate)
+
+	for _, key := range keysToUpdate {
+		curSw, newSw := currentMap[key], incomingMap[key]
+		if _, err := tx.ExecContext(ctx, stmt, newSw.LastOpenedAt, hostID, curSw.ID); err != nil {
```

Comment:
Going with a straightforward UPDATE for each software timestamp to update - typically, there should be a relatively low number of those per host (i.e. this is macos only, "user-facing apps" only, and users usually use a small subset of installed apps on a regular basis). There are ways to generate a batch-update, but they're all a bit hacky and unclear if they'd be significantly faster and generate less issues (plus, portability across versions and dbms implementations).